src/liballoc/collections/vec_deque.rs

```diff
@@ -1834,8 +1834,144 @@ impl<T> VecDeque<T> {
+                            copy_part(0, src_2, src_3, src, dst);
+                        }
+                        src_2 + src_3
+                    } else {
+                        //            src
+                        // [o o . . . . . . . o o o]
+                        //  2 2               1 1 1
+                        //
+                        //            dst
+                        // [. . . . . . . o o . . .]
+                        //  2 2 H             1 1 1
+                        let src_2 = src_total - src_before_wrap;
+
+                        unsafe {
+                            copy_part(src_start_1, dst_start_1, src_before_wrap, src, dst);
+                            copy_part(0, 0, src_2, src, dst);
+                        }
+                        src_2
+                    }
+                }
+                (false, true) => {
+                    //            src
+                    // [. . . o o o o o . . . .]
+                    //        1 1 2 2 2
+                    //
+                    //            dst
+                    // [. . . . . . . o o o . .]
+                    //  2 2 2 H             1 1
+                    let dst_1 = dst_wrap_point - dst_start_1;
+                    let src_start_2 = src_start_1 + dst_1;
+                    let dst_2 = src_total - dst_1;
+
+                    unsafe {
+                        copy_part(src_start_1, dst_start_1, dst_1, src, dst);
+                        copy_part(src_start_2, 0, dst_2, src, dst);
+                    }
+                    dst_2
+                }
+                (true, false) => {
+                    //            src
+                    // [o o . . . . . . . o o o]
+                    //  2 2               1 1 1
+                    //
+                    //            dst
+                    // [. o o . . . . . . . . .]
+                    //        1 1 1 2 2 H
+                    let src_1 = src_wrap_point - src_start_1;
+                    let dst_start_2 = dst_start_1 + src_1;
+                    let src_2 = src_total - src_1;
+
+                    unsafe {
+                        copy_part(src_start_1, dst_start_1, src_1, src, dst);
+                        copy_part(0, dst_start_2, src_2, src, dst);
+                    }
+                    dst_start_1 + src_1 + src_2
+                }
+                (false, false) => {
+                    //            src
+                    // [. . . o o o . . . . . .]
+                    //        1 1 1
+                    //
+                    //            dst
+                    // [. o o o o o . . . . . .]
+                    //              1 1 1 H
+                    unsafe {
+                        copy_part(src_start_1, dst_start_1, src_total, src, dst);
+                    }
+                    dst_start_1 + src_total
+                }
+            }
+        };
+        other.clear();
```

Comment:
Presumably, until this point we are in a Bad State where a given element actually exists twice, once in both collections, right? I believe that means the panic safety of this function needs to be documented to help prevent the next person to touching it from shooting themselves in the foot.